Add `config` command

abhi3700 opened the pull request:
## Description

This PR adds `config` command to the CLI tool that would allow user to change (if they wish for) the set params (during `init` command) like:
- [x] show already set config
- [x] switch chain
- [x] set farm-size
- [x] set reward-address
- [x] set node directory
- [x] set farm directory
- [x] set all params together

This also covers the issue: https://github.com/subspace/pulsar/issues/248#issue-1857779487

abhi3700 committed:
added boilerplate for 'config' command

abhi3700 committed:
Added logic for config function

abhi3700 committed:
added individual options possibility in config command

- wrapped config options with 'Option' type
- show config is working fine!
- added fn in utils
- prettify the config details

abhi3700 committed:
replaced bail w println because of better CLI interface that doesn not show debug info

abhi3700 committed:
changed the logic from 'set one param only' to 'either or all params'

abhi3700 committed:
config command modified

- removed redundant check for MIN_FARM_SIZE
- added config command in arrow_key_mode
- added usages except node, farm directory

abhi3700 committed:
Refactor utils and correct command order in interactive mode

- removed small fn from utils & used the code directly into placeholders
- corrected the order of commands in interactive mode

abhi3700 committed:
Manual testing of create_or_move_data fn, doc update for parameters

- thorough manual testing of `create_or_move_data` fn was done
- added usage for `node-dir` & `farm-dir` params & all params at a time.

nazar-pc commented on src/commands/config.rs:
```diff
@@ -0,0 +1,161 @@
+
+        // update (optional) the chain
+        if let Some(c) = chain {
+            config.chain = ChainConfig::from_str(&c)?;
```
Maybe check if the chain has changed before wiping it?

abhi3700 replied:
Although redundant.
[Commit](https://github.com/subspace/pulsar/commit/ef9072f468a1496d026ba1705289f26ca1533497).

nazar-pc reviewed (commented):
Please update against latest `main`

abhi3700 committed:
Merge branch 'main' of https://github.com/subspace/pulsar into add-config-cmd

# Conflicts:
#	Cargo.toml
#	pulsar/src/commands/config.rs

abhi3700 committed:
added missing pkg: serde_json

abhi3700 committed:
ensured check for already set chain, wipe after new chain set confirmation

abhi3700 committed:
fixed clippy errors

abhi3700 committed:
fixed 'cargo doc' issues

abhi3700 committed:
fixed build doc error

ParthDesai commented on pulsar/src/utils.rs:
```diff
@@ -388,3 +388,37 @@ impl<'de> Deserialize<'de> for Rewards {
+
+/// move data from old directory (if any) to new directory, or else just create
+/// the new directory.
+pub(crate) fn create_or_move_data(old_dir: PathBuf, new_dir: PathBuf) -> Result<()> {
```
Please write test for this method since if this method does not work properly node data could be corrupted or incomplete.

abhi3700 replied:
Added tests [here](https://github.com/subspace/pulsar/commit/9b2539e06ed38b12251ef7de1c25564b0090af32).

ParthDesai commented on pulsar/src/utils.rs:
```diff
@@ -388,3 +388,37 @@ impl<'de> Deserialize<'de> for Rewards {
+        }
+
+        // Check if the old directory is empty now and remove it
+        if fs::read_dir(&old_dir)?.next().is_none() {
```
In which scenario, the old directory is non-empty?

abhi3700 replied:
It’s possible that the data was manually deleted, perhaps by mistake. So, this is just a safety check.

ParthDesai replied:
Sorry, I did not get you. Can you elaborate?

abhi3700 replied:
Basically, if someone manually deletes by any chance, then the folder might be empty.
So, this is an additional check.

ParthDesai commented on pulsar/src/main.rs:
```diff
@@ -68,6 +69,23 @@ enum Commands {
+        about = "set the config params: chain, farm-size, reward-address, node-dir, farm-dir"
+    )]
+    Config {
+        #[arg(short, long, action)]
```
Would this be better UX?

To show config:
```
./pulsar config
```

To edit
```
./pulsar config --edit chain="new_chain" node_dir="new_dir"
```

ParthDesai commented on pulsar/src/commands/config.rs:
```diff
@@ -0,0 +1,168 @@
+        if let Some(fd) = farm_dir {
+            let farm_dir = PathBuf::from_str(&fd).expect("Invalid farm directory");
+            create_or_move_data(config.farmer.farm_directory.clone(), farm_dir.clone())?;
+            if farm_dir.exists() {
```
Why would it not exists? Since you are creating it just above.

abhi3700 replied:
I am able to realize after writing tests for `create_or_move_data` that it's redundant check.
Fixed in this [commit](https://github.com/subspace/pulsar/commit/96e408b430aa1b3316aa06762e2d88ced2f776fa).

abhi3700 committed:
Remove show flag from config command

`$ pulsar config -s` has been replaced with `$ pulsar config` to show the already set config params/fields details.

abhi3700 committed:
Correct rust doc for config

abhi3700 committed:
Add tests for fn used in config fn

abhi3700 committed:
Add tests

abhi3700 committed:
Updated tests for 'create_or_move_data' fn

abhi3700 committed:
removed redundant check

abhi3700 committed:
Refactor code with 'bail!'

ParthDesai commented on pulsar/src/tests.rs:
```diff
@@ -175,3 +177,48 @@ fn custom_log_dir_test() {
 }
+
+#[cfg(test)]
+mod create_or_move_data_tests {
```
You can just have `mod test`

abhi3700 replied:
Actually, the aim is to keep the related tests under one mod label.
Let me know your thoughts.

ParthDesai commented on pulsar/src/tests.rs:
```diff
@@ -175,3 +177,48 @@ fn custom_log_dir_test() {
+
+    /// Ensuring the function correctly handles the case where the old and new
+    /// directories are the same.
+    #[test]
```
I would prefer to test a scenario where we have actual files in old directory which we move into new and delete the old.

ParthDesai commented on pulsar/src/utils.rs:
```diff
@@ -388,3 +386,38 @@ impl<'de> Deserialize<'de> for Rewards {
+        for entry in entries {
+            let entry = entry?;
+            let file_name = entry.file_name();
+            fs::rename(entry.path(), new_dir.join(file_name))?;
```
What would happen if this fails? We are left with some data in old directory and some data in new one. Which means after this happen, user won't be able to use pulsar and have to fix the situation by hand.

I would like to have atomic move implemented here. Either we move everything or we move nothing.
